cdap-app-templates/cdap-program-report/src/main/java/co/cask/cdap/report/ReportGenerationSpark.java

```diff
@@ -141,17 +149,31 @@ public void getReports(HttpServiceRequest request, HttpServiceResponder responde
       responder.sendJson(200, reportList);
     }
 
+    private String getUserName(Map<String, List<String>> headers) {
+      String user = DEFAULT_USER_ID;
+      if (headers.containsKey(USER_ID) && headers.get(USER_ID).size() > 0) {
+        user = headers.get(USER_ID).get(0);
+      }
+      return user;
+    }
+
     /**
      * Gets a list of report statuses from all the unexpired reports starting from the given offset position,
      * with at most the limit number of report statuses in the list returned.
      *
      * @param offset the starting index position in the unexpired report directories
      * @param limit the maximum number of report statuses returned in the result
+     * @param user name of the user generating/requesting reports
      * @return a list of report statuses of unexpired reports
      * @throws Exception if getting status of a report fails
      */
-    private ReportList getReportList(int offset, int limit) throws Exception {
-      Location reportFilesetLocation = getDatasetBaseLocation(ReportGenerationApp.REPORT_FILESET);
+    private ReportList getReportList(int offset, int limit, String user) throws Exception {
+      Location reportFilesetLocation = getDatasetBaseLocation(ReportGenerationApp.REPORT_FILESET).append(user);
+      // we create directory for the user when the user generates a report first,
+      // return empty list if location doesn't exist.
+      if (!reportFilesetLocation.exists()) {
+        return new ReportList(offset, limit, 0, new ArrayList<>());
```

Comment:
Will the `ArrayList` getting modified later? If not, use `Collections.emptyList()` to create an immutable empty list

Reply:
yea, changed to emptyList